### DISCUSSION TITLE
Is the MV3 version of Privacy Badger entirely declarative?

### DISCUSSION THREAD

**Discussion by sillyjaybird** [EFForg/privacybadger#2992] [General]
I'm asking this question because I see on µBlock Origin Lite's GitHub site @gorhill says:

> What are the upsides of an MV3-based content blocker?
For Chromium-based browsers specifically, MV3-based content blockers will properly filter at browser launch, which is not the case for MV2-based content blockers. Note though that this is not an issue for Firefox MV2-based content blockers, see [uBlock Origin works best on Firefox / Browser launch](https://github.com/gorhill/uBlock/wiki/uBlock-Origin-works-best-on-Firefox#browser-launch).

[https://github.com/uBlockOrigin/uBOL-home/wiki/Frequently-asked-questions-(FAQ)#what-are-the-upsides-of-an-mv3-based-content-blocker](url)

>  uBO Lite being entirely declarative means it's service worker is lightweight and initialize optimally when activated. _Non-declarative MV3-based content blockers will suffer unreliable filtering when their service worker has been suspended, since waking up a service worker requires a lot of initialization work and delays time-critical filtering abilities._ Such content blocker may end up using trickery to force their service worker to always be up and running. 
>

[https://github.com/uBlockOrigin/uBOL-home/wiki/Frequently-asked-questions-(FAQ)#is-ubo-lite-a-bad-faith-attempt-at-converting-ubo-to-mv3](url)
More specifically, I'm wondering how local learning works in the MV3 version and whether or not it does so declaratively. I like the concept of Privacy Badger requiring little or no interaction from the user and the idea of using MV3's potential efficiency "advantage" for tracker blocking.

**Comment by ghostwords** [comment]
The MV3 version of Privacy Badger is [not Privacy Badger "lite"](https://github.com/EFForg/privacybadger/issues/2273#issuecomment-2142155316). Beyond that, I am not quite sure what exactly it is you're asking. What does "entirely declarative" mean?

**Reply by ghostwords** [reply]
Hmm, I *think* "entirely declarative" (from the [uBO Lite wiki](https://github.com/uBlockOrigin/uBOL-home/wiki/Frequently-asked-questions-(FAQ)#is-ubo-lite-a-bad-faith-attempt-at-converting-ubo-to-mv3) you link to) compares uBO Lite to other "full fat" MV3-based ad blockers as part of explaining the choices guiding the project.

OK, with that mind: Privacy Badger is not an ad blocker, and it doesn't use any ad blocker filter lists. So, while Privacy Badger is not "lite", Privacy Badger was never "full fat" in the first place. Also, while MV3 Privacy Badger should be ready to block trackers faster than before on startup, this wasn't a goal. The goal was to have the same functionality in MV3 as we do in MV2, and to have all features work as well as before as much as possible.

**Reply by sillyjaybird** [reply]
Thanks for your reply.
